File: api/app/data/fasting_observance_dates.py
"""Timing-correct viratham dates for 2026 (Nithra-style)."""

from __future__ import annotations

import re
from datetime import date


AMAVASAI_2026: set[date] = {
    date(2026, 1, 17),
    date(2026, 2, 16),
    date(2026, 2, 17),
    date(2026, 3, 18),
    date(2026, 4, 16),
    date(2026, 4, 17),
    date(2026, 5, 15),
    date(2026, 6, 14),
    date(2026, 7, 13),
    date(2026, 7, 14),
    date(2026, 8, 11),
    date(2026, 8, 12),
    date(2026, 9, 10),
    date(2026, 10, 9),
    date(2026, 10, 10),
    date(2026, 11, 8),
    date(2026, 11, 9),
    date(2026, 12, 7),
    date(2026, 12, 8),
}

POURNAMI_2026: set[date] = {
    date(2026, 1, 2),
    date(2026, 1, 3),
    date(2026, 1, 31),
    date(2026, 3, 2),
    date(2026, 3, 3),
    date(2026, 4, 1),
    date(2026, 4, 30),
    date(2026, 5, 1),
    date(2026, 5, 30),
    date(2026, 5, 31),
    date(2026, 6, 28),
    date(2026, 7, 28),
    date(2026, 7, 29),
    date(2026, 8, 27),
    date(2026, 9, 25),
    date(2026, 9, 26),
    date(2026, 10, 25),
    date(2026, 11, 23),
    date(2026, 11, 24),
    date(2026, 12, 23),
}

EKADASI_2026: set[date] = {
    date(2026, 1, 14),
    date(2026, 1, 29),
    date(2026, 2, 13),
    date(2026, 2, 27),
    date(2026, 3, 15),
    date(2026, 3, 29),
    date(2026, 4, 12),
    date(2026, 4, 27),
    date(2026, 5, 13),
    date(2026, 5, 27),
    date(2026, 6, 11),
    date(2026, 6, 25),
    date(2026, 7, 10),
    date(2026, 7, 25),
    date(2026, 8, 9),
    date(2026, 8, 22),
    date(2026, 9, 7),
    date(2026, 9, 22),
    date(2026, 10, 5),
    date(2026, 10, 22),
    date(2026, 11, 5),
    date(2026, 11, 21),
    date(2026, 12, 4),
    date(2026, 12, 20),
}

SASHTI_2026: set[date] = {
    date(2026, 1, 9),
    date(2026, 1, 23),
    date(2026, 2, 6),
    date(2026, 2, 23),
    date(2026, 3, 9),
    date(2026, 3, 24),
    date(2026, 4, 8),
    date(2026, 4, 22),
    date(2026, 5, 8),
    date(2026, 5, 22),
    date(2026, 6, 5),
    date(2026, 6, 20),
    date(2026, 7, 6),
    date(2026, 7, 18),
    date(2026, 8, 4),
    date(2026, 8, 18),
    date(2026, 9, 2),
    date(2026, 9, 17),
    date(2026, 10, 2),
    date(2026, 10, 15),
    date(2026, 10, 31),
    date(2026, 11, 14),
    date(2026, 11, 28),
    date(2026, 12, 15),
    date(2026, 12, 29),
}

PRADOSHAM_2026: set[date] = {
    date(2026, 1, 1),
    date(2026, 1, 16),
    date(2026, 1, 31),
    date(2026, 2, 15),
    date(2026, 3, 1),
    date(2026, 3, 17),
    date(2026, 3, 31),
    date(2026, 4, 15),
    date(2026, 4, 29),
    date(2026, 5, 15),
    date(2026, 5, 29),
    date(2026, 6, 13),
    date(2026, 6, 26),
    date(2026, 7, 12),
    date(2026, 7, 27),
    date(2026, 8, 10),
    date(2026, 8, 26),
    date(2026, 9, 9),
    date(2026, 9, 24),
    date(2026, 10, 8),
    date(2026, 10, 24),
    date(2026, 11, 7),
    date(2026, 11, 21),
    date(2026, 12, 5),
    date(2026, 12, 22),
}

SIVARATRI_2026: set[date] = {
    date(2026, 1, 16),
    date(2026, 2, 15),
    date(2026, 3, 17),
    date(2026, 4, 15),
    date(2026, 5, 15),
    date(2026, 6, 13),
    date(2026, 7, 12),
    date(2026, 8, 10),
    date(2026, 9, 9),
    date(2026, 10, 8),
    date(2026, 11, 7),
    date(2026, 12, 6),
}

CHATURTHI_2026: set[date] = {
    date(2026, 1, 21),
    date(2026, 2, 21),
    date(2026, 3, 22),
    date(2026, 4, 20),
    date(2026, 5, 20),
    date(2026, 6, 18),
    date(2026, 7, 17),
    date(2026, 8, 16),
    date(2026, 9, 15),
    date(2026, 10, 13),
    date(2026, 11, 13),
    date(2026, 12, 13),
}

SANKATAHARA_2026: set[date] = {
    date(2026, 1, 7),
    date(2026, 2, 4),
    date(2026, 3, 7),
    date(2026, 4, 6),
    date(2026, 5, 6),
    date(2026, 6, 4),
    date(2026, 7, 4),
    date(2026, 8, 2),
    date(2026, 9, 1),
    date(2026, 9, 30),
    date(2026, 10, 29),
    date(2026, 11, 28),
    date(2026, 12, 27),
}

KIRUTHIGAI_2026: set[date] = {
    date(2026, 1, 28),
    date(2026, 2, 24),
    date(2026, 3, 23),
    date(2026, 5, 17),
    date(2026, 6, 13),
    date(2026, 7, 11),
    date(2026, 8, 7),
    date(2026, 9, 3),
    date(2026, 10, 1),
    date(2026, 10, 28),
    date(2026, 11, 24),
    date(2026, 12, 22),
}

THIRUVONAM_2026: set[date] = {
    date(2026, 1, 20),
    date(2026, 2, 16),
    date(2026, 3, 15),
    date(2026, 4, 12),
    date(2026, 5, 9),
    date(2026, 6, 5),
    date(2026, 6, 6),
    date(2026, 7, 3),
    date(2026, 7, 30),
    date(2026, 8, 26),
    date(2026, 9, 23),
    date(2026, 10, 20),
    date(2026, 11, 16),
    date(2026, 12, 13),
    date(2026, 12, 14),
}


_CHATOORTHI_RE = re.compile(r"சதுர்த்தி(?!சி)")


def sunrise_tithi_text(tithi_text: str) -> str:
    if "பின்பு" in tithi_text:
        return tithi_text.split("பின்பு", 1)[0].strip()
    return tithi_text.strip()


def after_pinbu_text(tithi_text: str) -> str:
    if "பின்பு" in tithi_text:
        return tithi_text.split("பின்பு", 1)[1].strip()
    return ""


def day_has_amavasai(tithi_text: str) -> bool:
    return "அமாவாசை" in tithi_text


def day_has_pournami(tithi_text: str) -> bool:
    return "பௌர்ணமி" in tithi_text or "பவுர்ணமி" in tithi_text


def day_has_ekadasi(tithi_text: str) -> bool:
    return "ஏகாதசி" in tithi_text


def day_has_sashti(tithi_text: str) -> bool:
    return "சஷ்டி" in tithi_text


def day_has_pradosham(tithi_text: str) -> bool:
    return "திரயோதசி" in tithi_text


def day_has_sivaratri(tithi_text: str) -> bool:
    aft = after_pinbu_text(tithi_text)
    if aft:
        return "சதுர்த்தசி" in aft and "தேய்பிறை" in aft
    return "சதுர்த்தசி" in tithi_text and "தேய்பிறை" in tithi_text


def day_has_chaturthi(tithi_text: str) -> bool:
    for part in tithi_text.split("பின்பு"):
        if _CHATOORTHI_RE.search(part) and "வளர்பிறை" in part:
            return True
    return False


def day_has_sankatahara(tithi_text: str) -> bool:
    for part in tithi_text.split("பின்பு"):
        if _CHATOORTHI_RE.search(part) and "தேய்பிறை" in part:
            return True
    return False


def day_has_kiruthigai(nak_text: str) -> bool:
    return "கிருத்திகை" in nak_text


def day_has_thiruvonam(nak_text: str) -> bool:
    return "திருவோணம்" in nak_text


def is_amavasai_date(on_date: date) -> bool:
    return on_date in AMAVASAI_2026


def is_pournami_date(on_date: date) -> bool:
    return on_date in POURNAMI_2026


def is_sivaratri_date(on_date: date) -> bool:
    return on_date in SIVARATRI_2026


def sunrise_has_amavasai(tithi_text: str) -> bool:
    return day_has_amavasai(tithi_text)


def sunrise_has_pournami(tithi_text: str) -> bool:
    return day_has_pournami(tithi_text)

File: app/lib/data/fasting_observance_dates.dart
/// Viratham dates for 2026 — timing-correct (Nithra-style).
///
/// Moon (அமாவாசை / பௌர்ணமி): mark a day if the tithi is present at sunset,
/// begins that calendar day, OR is present at sunrise and continues past noon
/// (drops short morning-only remnants like Aug 28 / Sep 11).
/// Other tithi virathams: last sunrise day per occurrence.
/// சிவராத்திரி: day தேய்பிறை சதுர்த்தசி begins (after பின்பு).
/// Nakshatra days: present in sunrise nakshatra text.
library;


const amavasai2026 = {
  '2026-01-17',
  '2026-02-16',
  '2026-02-17',
  '2026-03-18',
  '2026-04-16',
  '2026-04-17',
  '2026-05-15',
  '2026-06-14',
  '2026-07-13',
  '2026-07-14',
  '2026-08-11',
  '2026-08-12',
  '2026-09-10',
  '2026-10-09',
  '2026-10-10',
  '2026-11-08',
  '2026-11-09',
  '2026-12-07',
  '2026-12-08',
};


const pournami2026 = {
  '2026-01-02',
  '2026-01-03',
  '2026-01-31',
  '2026-03-02',
  '2026-03-03',
  '2026-04-01',
  '2026-04-30',
  '2026-05-01',
  '2026-05-30',
  '2026-05-31',
  '2026-06-28',
  '2026-07-28',
  '2026-07-29',
  '2026-08-27',
  '2026-09-25',
  '2026-09-26',
  '2026-10-25',
  '2026-11-23',
  '2026-11-24',
  '2026-12-23',
};


const ekadasi2026 = {
  '2026-01-14',
  '2026-01-29',
  '2026-02-13',
  '2026-02-27',
  '2026-03-15',
  '2026-03-29',
  '2026-04-12',
  '2026-04-27',
  '2026-05-13',
  '2026-05-27',
  '2026-06-11',
  '2026-06-25',
  '2026-07-10',
  '2026-07-25',
  '2026-08-09',
  '2026-08-22',
  '2026-09-07',
  '2026-09-22',
  '2026-10-05',
  '2026-10-22',
  '2026-11-05',
  '2026-11-21',
  '2026-12-04',
  '2026-12-20',
};


const sashti2026 = {
  '2026-01-09',
  '2026-01-23',
  '2026-02-06',
  '2026-02-23',
  '2026-03-09',
  '2026-03-24',
  '2026-04-08',
  '2026-04-22',
  '2026-05-08',
  '2026-05-22',
  '2026-06-05',
  '2026-06-20',
  '2026-07-06',
  '2026-07-18',
  '2026-08-04',
  '2026-08-18',
  '2026-09-02',
  '2026-09-17',
  '2026-10-02',
  '2026-10-15',
  '2026-10-31',
  '2026-11-14',
  '2026-11-28',
  '2026-12-15',
  '2026-12-29',
};


const pradosham2026 = {
  '2026-01-01',
  '2026-01-16',
  '2026-01-31',
  '2026-02-15',
  '2026-03-01',
  '2026-03-17',
  '2026-03-31',
  '2026-04-15',
  '2026-04-29',
  '2026-05-15',
  '2026-05-29',
  '2026-06-13',
  '2026-06-26',
  '2026-07-12',
  '2026-07-27',
  '2026-08-10',
  '2026-08-26',
  '2026-09-09',
  '2026-09-24',
  '2026-10-08',
  '2026-10-24',
  '2026-11-07',
  '2026-11-21',
  '2026-12-05',
  '2026-12-22',
};


const sivaratri2026 = {
  '2026-01-16',
  '2026-02-15',
  '2026-03-17',
  '2026-04-15',
  '2026-05-15',
  '2026-06-13',
  '2026-07-12',
  '2026-08-10',
  '2026-09-09',
  '2026-10-08',
  '2026-11-07',
  '2026-12-06',
};


const chaturthi2026 = {
  '2026-01-21',
  '2026-02-21',
  '2026-03-22',
  '2026-04-20',
  '2026-05-20',
  '2026-06-18',
  '2026-07-17',
  '2026-08-16',
  '2026-09-15',
  '2026-10-13',
  '2026-11-13',
  '2026-12-13',
};


const sankatahara2026 = {
  '2026-01-07',
  '2026-02-04',
  '2026-03-07',
  '2026-04-06',
  '2026-05-06',
  '2026-06-04',
  '2026-07-04',
  '2026-08-02',
  '2026-09-01',
  '2026-09-30',
  '2026-10-29',
  '2026-11-28',
  '2026-12-27',
};


const kiruthigai2026 = {
  '2026-01-28',
  '2026-02-24',
  '2026-03-23',
  '2026-05-17',
  '2026-06-13',
  '2026-07-11',
  '2026-08-07',
  '2026-09-03',
  '2026-10-01',
  '2026-10-28',
  '2026-11-24',
  '2026-12-22',
};


const thiruvonam2026 = {
  '2026-01-20',
  '2026-02-16',
  '2026-03-15',
  '2026-04-12',
  '2026-05-09',
  '2026-06-05',
  '2026-06-06',
  '2026-07-03',
  '2026-07-30',
  '2026-08-26',
  '2026-09-23',
  '2026-10-20',
  '2026-11-16',
  '2026-12-13',
  '2026-12-14',
};



bool _has(Set<String> s, String dateKey) => s.contains(dateKey);

bool isAmavasaiDate(String dateKey) => _has(amavasai2026, dateKey);
bool isPournamiDate(String dateKey) => _has(pournami2026, dateKey);
bool isEkadasiDate(String dateKey) => _has(ekadasi2026, dateKey);
bool isSashtiDate(String dateKey) => _has(sashti2026, dateKey);
bool isPradoshamDate(String dateKey) => _has(pradosham2026, dateKey);
bool isSivaratriDate(String dateKey) => _has(sivaratri2026, dateKey);
bool isChaturthiDate(String dateKey) => _has(chaturthi2026, dateKey);
bool isSankataharaDate(String dateKey) => _has(sankatahara2026, dateKey);
bool isKiruthigaiDate(String dateKey) => _has(kiruthigai2026, dateKey);
bool isThiruvonamDate(String dateKey) => _has(thiruvonam2026, dateKey);

String sunriseTithiText(String tithiText) {
  final idx = tithiText.indexOf('பின்பு');
  if (idx < 0) return tithiText.trim();
  return tithiText.substring(0, idx).trim();
}

String afterPinbuText(String tithiText) {
  final idx = tithiText.indexOf('பின்பு');
  if (idx < 0) return '';
  return tithiText.substring(idx + 'பின்பு'.length).trim();
}

bool dayHasAmavasai(String tithiText) => tithiText.contains('அமாவாசை');
bool dayHasPournami(String tithiText) =>
    tithiText.contains('பௌர்ணமி') || tithiText.contains('பவுர்ணமி');
bool dayHasEkadasi(String tithiText) => tithiText.contains('ஏகாதசி');
bool dayHasSashti(String tithiText) => tithiText.contains('சஷ்டி');
bool dayHasPradosham(String tithiText) => tithiText.contains('திரயோதசி');
bool dayHasKiruthigai(String nakText) => nakText.contains('கிருத்திகை');
bool dayHasThiruvonam(String nakText) => nakText.contains('திருவோணம்');

final _chaturthiRe = RegExp(r'சதுர்த்தி(?!சி)');

bool dayHasChaturthi(String tithiText) {
  for (final part in tithiText.split('பின்பு')) {
    if (_chaturthiRe.hasMatch(part) && part.contains('வளர்பிறை')) return true;
  }
  return false;
}

bool dayHasSankatahara(String tithiText) {
  for (final part in tithiText.split('பின்பு')) {
    if (_chaturthiRe.hasMatch(part) && part.contains('தேய்பிறை')) return true;
  }
  return false;
}

bool dayHasSivaratri(String tithiText) {
  final aft = afterPinbuText(tithiText);
  if (aft.isNotEmpty) {
    return aft.contains('சதுர்த்தசி') && aft.contains('தேய்பிறை');
  }
  return tithiText.contains('சதுர்த்தசி') && tithiText.contains('தேய்பிறை');
}
